skimage/morphology/_max_tree.pyx

```diff
@@ -0,0 +1,459 @@
+ctypedef np.int32_t DTYPE_INT32_t
+ctypedef np.uint32_t DTYPE_UINT32_t
+ctypedef np.uint64_t DTYPE_UINT64_t
+ctypedef np.int64_t DTYPE_INT64_t
+ctypedef np.uint8_t DTYPE_BOOL_t
+ctypedef np.uint8_t DTYPE_UINT8_t
+ctypedef np.int8_t DTYPE_INT8_t
+
+ctypedef fused dtype_t:
+    np.uint8_t
+    np.uint16_t
+    np.uint32_t
+    np.uint64_t
+    np.int8_t
+    np.int16_t
+    np.int32_t
+    np.int64_t
+    np.float32_t
+    np.float64_t
+
+cdef inline DTYPE_UINT64_t uint64_max(DTYPE_UINT64_t a, DTYPE_UINT64_t b): return a if a >= b else b
+cdef inline DTYPE_UINT64_t uint64_min(DTYPE_UINT64_t a, DTYPE_UINT64_t b): return a if a <= b else b
+
+
+# recursive function to get the root of the current tree. Importantly, the
+# function changes the tree (path compression), that reduces the complexity
+# from O(n*n) to O(n*log(n)).
+cdef DTYPE_INT64_t find_root(DTYPE_INT64_t[::1] parent, DTYPE_INT64_t index):
+    if parent[index] != index:
+        parent[index] = find_root(parent, parent[index])
+    return parent[index]
+
+# canonicalize generates a max-tree for which every node's
+# parent is a canonical node. Namely, either the representative node of the
+# same level or the representative node at the level below.
+cdef void canonize(dtype_t[::1] image, DTYPE_INT64_t[::1] parent,
+                   DTYPE_INT64_t[::1] sorted_indices):
+    cdef DTYPE_INT64_t q = 0
+    for p in sorted_indices:
+        q = parent[p]
+        if image[q] == image[parent[q]]:
+            parent[p] = parent[q]
+    return
+
+
+# helper function: a list of offsets is transformed to a list of points,
+# which are stored in a numpy array: rows correspond to different points,
+# columns to the dimensions (typically 2 or 3, but there is no limitation).
+cdef np.ndarray[DTYPE_INT32_t, ndim = 2] offsets_to_points(DTYPE_INT32_t[::1] offsets,
+                                                           DTYPE_INT32_t[::1] shape):
+    cdef DTYPE_INT32_t number_of_dimensions = len(shape)
+    cdef DTYPE_INT32_t number_of_points = len(offsets)
+    cdef np.ndarray[DTYPE_INT32_t, ndim = 2] points = np.zeros((number_of_points,
+                                                                number_of_dimensions),
+                                                               dtype=np.int32)
+    cdef DTYPE_INT32_t neg_shift = - np.min(offsets)
+
+    cdef DTYPE_INT32_t i, offset, curr_index, coord
+
+    center_point = np.unravel_index(neg_shift, shape)
+
+    for i, offset in enumerate(offsets):
+        current_point = np.unravel_index(offset + neg_shift, shape)
+        for d in range(number_of_dimensions):
+            points[i, d] = current_point[d] - center_point[d]
+
+    return points
+
+
+# checks whether a neighbor of a given pixel is inside the image plane.
+# The pixel is given in form of an index of a raveled array, the offset
+# is given as a list of coordinates. This function is relatively time-consuming
+# and should only be applied to border-pixels (defined by mask).
```

Comment:
I'm also not sure what this is doing. To me it sounds like it's figuring out whether this is a border pixel, but you already have that information since you say you should only apply this to border pixels?

Reply:
I rewrote the doc and renamed the function. 
This function tests for a given border pixel and an offset, whether the resulting neighbour (pixel + offset) is still inside the image. 

But I agree: this is certainly not the nicest way of doing this ...